opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -208,8 +209,38 @@ def set_attribute(self, key: str, value: types.AttributeValue) -> None:
         if has_ended:
             logger.warning("Setting attribute on ended span.")
             return
+
+        if not isinstance(value, (int, float, bool, str, list, tuple)):
+            logger.warning("invalid type for attribute value")
+            return
+        if isinstance(value, (list, tuple)) and len(value) > 0:
+            return_code = self._check_sequence(value)
+            if return_code is not None:
+                logger.warning("%s in attribute value sequence", return_code)
+                return
+
         self.attributes[key] = value
 
+    @staticmethod
+    def _check_sequence(sequence: (list, tuple)) -> Optional[str]:
+        """
+        Checks if sequence items are valid and identical
+        """
+        first_element_type = type(sequence[0])
+
+        if issubclass(type(sequence[0]), Number):
+            first_element_type = Number
+
+        for element in sequence:
+
+            if not isinstance(element, (bool, str, Number)):
```

Comment:
I feel like this block could be optimized a little bit. ultimately, you're validating the intersection of two requirements:

1. all elements of the sequence are the same type
2. the element is one of (bool, str, Number).

First step I see is to check the first element, and see if it's a valid type. after that, you can be assured that no element can be an invalid type, as in that case it would fail the type mismatch check. 

So ultimately your logic could be rewritten to:

1. check if the first element type is invalid
2. check all remaining elements to see if it does not match the first element type.

Not a huge optimization, but saves an isinstance check over a tuple over and over again.

Reply:
@jakemalachowski implemented like that before, @toumorokoshi. I suggested this approach because it makes the error reporting more accurate. If the approach above is followed then a invalid-typed element being present in the first position will be reported as an invalid-typed element being present in the sequence. But if the invalid-typed element is present in any other position of the sequence, then it will be reported as an element that does not match the type of the rest of the elements. I find that to be inconsistent, to have a invalid-typed element being reported in two different ways depending on its position in the sequence.